homeassistant/components/update/__init__.py

```diff
@@ -0,0 +1,253 @@
+INFO_CALLBACK_TIMEOUT = 5
+STORAGE_VERSION = 1
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the Update integration."""
+    hass.data[DOMAIN] = UpdateManager(hass=hass)
+
+    websocket_api.async_register_command(hass, handle_info)
+    websocket_api.async_register_command(hass, handle_update)
+    websocket_api.async_register_command(hass, handle_skip)
+
+    await integration_platform.async_process_integration_platforms(
+        hass, DOMAIN, _register_update_platform
+    )
+
+    return True
+
+
+async def _register_update_platform(
+    hass: HomeAssistant,
+    integration_domain: str,
+    platform: UpdatePlatformProtocol,
+) -> None:
+    """Register an update platform."""
+    manager: UpdateManager = hass.data[DOMAIN]
+    manager.add_platform(integration_domain, platform)
+
+
+@websocket_api.websocket_command({vol.Required("type"): "update/info"})
+@websocket_api.async_response
+async def handle_info(
+    hass: HomeAssistant,
+    connection: websocket_api.ActiveConnection,
+    msg: dict[str, Any],
+) -> None:
+    """Get pending updates from all platforms."""
+    manager: UpdateManager = hass.data[DOMAIN]
+    updates = await manager.gather_updates()
+    connection.send_result(msg["id"], updates)
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "update/skip",
+        vol.Required("domain"): str,
+        vol.Required("identifier"): str,
+        vol.Required("version"): str,
+    }
+)
+@callback
+def handle_skip(
+    hass: HomeAssistant,
+    connection: websocket_api.ActiveConnection,
+    msg: dict[str, Any],
+) -> None:
+    """Skip an update."""
+    manager: UpdateManager = hass.data[DOMAIN]
+
+    if not manager.domain_is_valid(msg["domain"]):
+        connection.send_error(
+            msg["id"], websocket_api.ERR_NOT_FOUND, "Domain not supported"
+        )
+        return
+
+    manager.skip_update(msg["domain"], msg["identifier"], msg["version"])
+    connection.send_result(msg["id"])
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "update/update",
+        vol.Required("domain"): str,
+        vol.Required("identifier"): str,
+        vol.Required("version"): str,
+        vol.Optional("backup"): bool,
+    }
+)
+@websocket_api.async_response
+async def handle_update(
+    hass: HomeAssistant,
+    connection: websocket_api.ActiveConnection,
+    msg: dict[str, Any],
+) -> None:
+    """Handle an update."""
+    manager: UpdateManager = hass.data[DOMAIN]
+
+    if not manager.domain_is_valid(msg["domain"]):
+        connection.send_error(
+            msg["id"], websocket_api.ERR_NOT_FOUND, "Domain not supported"
+        )
+        return
+
+    if not await manager.perform_update(
+        domain=msg["domain"],
+        identifier=msg["identifier"],
+        version=msg["version"],
+        backup=msg.get("backup"),
+    ):
+        connection.send_error(msg["id"], "update_failed", "Update failed")
```

Comment:
Should we support an exception to indicate why an update failed so we can add a better error message beyond "Update failed"